src/coreclr/vm/peassembly.cpp

```diff
@@ -359,19 +395,33 @@ void PEAssembly::ConvertMDInternalToReadWrite()
     // Swap the pointers in a thread safe manner.  If the contents of *ppImport
     //  equals pOld then no other thread got here first, and the old contents are
     //  replaced with pNew.  The old contents are returned.
-    if (InterlockedCompareExchangeT(&m_pMDImport, pNew, pOld) == pOld)
-    {
-        //if the debugger queries, it will now see that we have RW metadata
-        m_MDImportIsRW_Debugger_Use_Only = TRUE;
 
-        // Swapped -- get the metadata to hang onto the old Internal import.
-        HRESULT hr=m_pMDImport->SetUserContextData(pOld);
-        _ASSERTE(SUCCEEDED(hr)||!"Leaking old MDImport");
-        IfFailThrow(hr);
-    }
-    else
-    {   // Some other thread finished first.  Just free the results of this conversion.
+    if (InterlockedCompareExchangeT(&m_pConvertedMDImport, pNew, NULL) != NULL)
+    {
+        // If we lost the race, free our copy of the RW MD. However, we might still
+        // be the only requesting swapping and our caller would expect m_pMDImport to
+        // be RW. Try to set it to the converted import.
         pNew->Release();
+        pNew = m_pConvertedMDImport;
+    }
+
+    if (openForWriting)
+    {
+        if (InterlockedCompareExchangeT(&m_pMDImport, pNew, pOld) == pOld)
+        {
+            //if the debugger queries, it will now see that we have RW metadata
+            m_MDImportIsRW_Debugger_Use_Only = TRUE;
+
+            // Swapped -- get the metadata to hang onto the old Internal import.
+            // Additionally, now both m_pMDImport and m_pConvertedMDImport point to
+            // the RW MDImport. Up the references such that both have to be freed
+            // before freeing the RW copy and the RO that hangs off it.
+            pNew->AddRef();
```

Comment:
The `pNew` was returned from above with an `AddRef()`, see `GetMetaDataInternalInterfaceFromPublic()`, why is this neded here? I think I get the case if `pNew = m_pConvertedMDImport;` is done, but in that case I would expect an `pNew->AddRef()` there as opposed to here.